lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,241 @@
+        add_card(post, options)
+
+        commit('sale', post, request_headers(options))
+      end
+
+      def authorize(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_card(post, options)
+        add_customer(post, options)
+
+        commit('authorization', post, request_headers(options))
+      end
+
+      def capture(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_card(post, options)
+
+        commit('capture', post, request_headers(options))
+      end
+
+      def refund(money, options = {})
+        post = {}
+        add_datetime(post, options)
+        api_options(post, options)
+        add_invoice(post, money, options)
+        add_clerk(post, money, options)
+        add_transaction(post, options)
+        add_device(post, options)
+        add_p2pe(post, options)
+        add_emv(post, options)
+        add_customer(post, options)
+        add_card(post, options)
+
+        commit('refund', post, request_headers(options))
+      end
+
+      def void(options = {})
+        post = {}
+        commit('invoice', post, request_headers(options))
+      end
+
+      def verify(credit_card, options = {})
+        MultiResponse.run(:use_first_response) do |r|
+          r.process { authorize(100, credit_card, options) }
+          r.process(:ignore_result) { void(r.authorization, options) }
+        end
+      end
+
+      private
+
+      def add_clerk(post, money, options)
+        post[:clerk] = {}
+        post[:clerk][:numericId] = options[:clerk][:numeric_id]
+      end
+
+      def add_invoice(post, money, options)
+        post[:amount] = {}
+        post[:amount][:total] = money
+        post[:amount][:tax] = options[:amount][:tax] || 0.0
+      end
+
+      def api_options(post, options)
+        post[:apiOptions] = options[:api_options] if options[:api_options].present?
+      end
+
+      def add_credentials(post, options)
+        post[:dateTime] = options[:date_time] || DateTime.now.strftime('%Y-%m-%dT%H:%M:%S.%N+%H:%M')
+        post[:credential] = {}
+        post[:credential][:clientGuid] = options[:credential][:client_guid]
+        post[:credential][:authToken] = options[:credential][:auth_token]
+      end
+
+      def add_datetime(post, options)
+        post[:dateTime] = options[:dateTime] || DateTime.now.strftime('%Y-%m-%dT%H:%M:%S.%N+%H:%M')
+      end
+
+      def add_transaction(post, options)
+        post[:transaction] = {}
+        post[:transaction][:invoice] = options[:transaction][:invoice]
+        post[:transaction][:notes] = options[:transaction][:notes] if options[:transaction][:notes].present?
+        if options[:transaction][:purchase_card].present?
```

Comment:
I think in the case of `purchase_card` it's fine to have the child options, but we should do away with the `transaction` nesting.

Reply:
Done